precompiles/evidence/evidence.go

```diff
@@ -0,0 +1,149 @@
+
+import (
+	"embed"
+	"fmt"
+
+	evidencekeeper "cosmossdk.io/x/evidence/keeper"
+
+	"cosmossdk.io/log"
+	storetypes "cosmossdk.io/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authzkeeper "github.com/cosmos/cosmos-sdk/x/authz/keeper"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+
+	cmn "github.com/evmos/evmos/v20/precompiles/common"
+	"github.com/evmos/evmos/v20/x/evm/core/vm"
+	evmtypes "github.com/evmos/evmos/v20/x/evm/types"
+)
+
+var _ vm.PrecompiledContract = &Precompile{}
+
+// Embed abi json file to the executable binary. Needed when importing as dependency.
+//
+//go:embed abi.json
+var f embed.FS
+
+// Precompile defines the precompiled contract for evidence.
+type Precompile struct {
+	cmn.Precompile
+	evidenceKeeper evidencekeeper.Keeper
+}
+
+// LoadABI loads the evidence ABI from the embedded abi.json file
+// for the evidence precompile.
+func LoadABI() (abi.ABI, error) {
+	return cmn.LoadABI(f, "abi.json")
+}
+
+// NewPrecompile creates a new evidence Precompile instance as a
+// PrecompiledContract interface.
+func NewPrecompile(
+	evidenceKeeper evidencekeeper.Keeper,
+	authzKeeper authzkeeper.Keeper,
+) (*Precompile, error) {
+	abi, err := LoadABI()
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Precompile{
+		Precompile: cmn.Precompile{
+			ABI:                  abi,
+			AuthzKeeper:          authzKeeper,
+			KvGasConfig:          storetypes.KVGasConfig(),
+			TransientKVGasConfig: storetypes.TransientGasConfig(),
+			ApprovalExpiration:   cmn.DefaultExpirationDuration, // should be configurable in the future.
+		},
+		evidenceKeeper: evidenceKeeper,
+	}
+
+	// SetAddress defines the address of the evidence precompiled contract.
+	p.SetAddress(common.HexToAddress(evmtypes.EvidencePrecompileAddress))
+
+	return p, nil
+}
+
+// RequiredGas calculates the precompiled contract's base gas rate.
+func (p Precompile) RequiredGas(input []byte) uint64 {
+	// NOTE: This check avoid panicking when trying to decode the method ID
+	if len(input) < 4 {
+		return 0
+	}
+	methodID := input[:4]
+
+	method, err := p.MethodById(methodID)
+	if err != nil {
+		// This should never happen since this method is going to fail during Run
+		return 0
+	}
+
+	return p.Precompile.RequiredGas(input, p.IsTransaction(method))
+}
+
+// Run executes the precompiled contract evidence methods defined in the ABI.
+func (p Precompile) Run(evm *vm.EVM, contract *vm.Contract, readOnly bool) (bz []byte, err error) {
+	ctx, stateDB, snapshot, method, initialGas, args, err := p.RunSetup(evm, contract, readOnly, p.IsTransaction)
+	if err != nil {
+		return nil, err
+	}
+
+	// This handles any out of gas errors that may occur during the execution of a precompile tx or query.
+	// It avoids panics and returns the out of gas error so the EVM can continue gracefully.
+	defer cmn.HandleGasError(ctx, contract, initialGas, &err)()
+
+	if err := stateDB.Commit(); err != nil {
+		return nil, err
+	}
```

Comment:
I believe this should be removed. All updates are tracked with the journal entries at line 126 and committed only once in the [state transition file](https://github.com/evmos/evmos/blob/ad1f0979452e3b4dfae4fe91fcc9a2f09c18bd9d/x/evm/keeper/state_transition.go#L351-L355).

Do you confirm @GAtom22 ?

Reply:
The commit happens here:
https://github.com/evmos/evmos/blob/ef391ddb05588202ed9e7cffdd0790271a7f1633/precompiles/common/precompile.go#L97-L99